File: src/main/scala/edu/stanford/w1/MergeSort.scala
//package edu.stanford.w1
//
//import scala.collection.mutable
//
//object MergeSort {
//  def mergeSort[T <: Ordered[T]](array: Seq[T]): Seq[T] = {
//    new MergeSort[T]().sort(array)
//  }
//}
//
//class MergeSort[T <: Ordered[T]] {
//
//  private def sort(array: Seq[T]): Seq[T] = {
//    val len = array.length
//    if (len < 2) {
//      return array
//    }
//    val mid = len / 2
//    merge(sort(array.slice(0, mid)), sort(array.slice(mid, len)))
//  }
//
//  private def merge(left: Seq[T], right: Seq[T]): Seq[T] = {
//
//    val mLen = left.length + right.length
//
//    val merged = mutable.IndexedSeq[T]()
//
//    var k = 0
//    var i = 0
//    var j = 0
//
//    while (k < mLen) {
//
//      if (i >= left.length) {
//        return merged ++ right.slice(j, right.length)
//      } else if (j >= right.length) {
//        return merged ++ left.slice(i, left.length)
//      }
//
//      if (left(i) < right(j)) {
//        merged(k) = left(i)
//        i += 1
//      } else {
//        merged(k) = right(j)
//        j += 1
//      }
//
//      k += 1
//    }
//
//    merged
//  }
//
//}

File: src/main/java/edu/stanford/w1/MergeSort.java
package edu.stanford.w1;

import java.lang.reflect.Array;

import static edu.stanford.common.CommonUtils.addAll;
import static java.util.Arrays.copyOfRange;

public class MergeSort<T extends Comparable<T>> {

    private T[] sort(T[] array) {
        int len = array.length;

        if (len < 2) {
            return array;
        }
        final int mid = len / 2;

        return merge(
                sort(copyOfRange(array, 0, mid)),
                sort(copyOfRange(array, mid, len))
        );
    }

    private T[] merge(T[] left, T[] right) {
        int mLen = left.length + right.length;
        @SuppressWarnings("unchecked")
        T[] merged = (T[]) Array.newInstance(left.getClass().getComponentType(), mLen);

        for (int k = 0, i = 0, j = 0; k < mLen; k++) {
            if (i >= left.length) {
                addAll(merged, right, k, j);
                break;
            } else if (j >= right.length) {
                addAll(merged, left, k, i);
                break;
            }

            if (left[i].compareTo(right[j]) < 0) {
                merged[k] = left[i];
                i++;
            } else {
                merged[k] = right[j];
                j++;
            }
        }
        return merged;
    }

    public static <T extends Comparable<T>> T[] mergeSort(T[] array) {
        return new MergeSort<T>().sort(array);
    }

}
